src/ui/public/timepicker/timepicker.js

```diff
@@ -78,36 +97,46 @@ module.directive('kbnTimepicker', function (quickRanges, timeUnits, refreshInter
           case 'quick':
             break;
           case 'relative':
-            let fromParts = $scope.from.toString().split('-');
-            let relativeParts = [];
+            var durationTo = moment.duration(moment().diff(dateMath.parse($scope.to)));
+            var durationFrom = moment.duration(moment().diff(dateMath.parse($scope.from)));
+
+            var unitsTo = _.pluck(_.clone($scope.relativeOptions.to).reverse(), 'value');
+            var unitsFrom = _.pluck(_.clone($scope.relativeOptions.from).reverse(), 'value');
+
+            if ($scope.from.toString().split('/')[1]) $scope.relative.from.round = true;
+            if ($scope.to.toString().split('/')[1]) $scope.relative.to.round = true;
 
-            // Try to parse the relative time, if we can't use moment duration to guestimate
-            if ($scope.to.toString() === 'now' && fromParts[0] === 'now' && fromParts[1]) {
-              relativeParts = fromParts[1].match(/([0-9]+)([smhdwMy]).*/);
+            var i;
+
+            for (i = 0; i < unitsTo.length; i++) {
+              var asTo = durationTo.as(unitsTo[i]);
+
+              if (asTo > 1) {
+                $scope.relative.to.count = Math.round(asTo);
+                $scope.relative.to.unit = unitsTo[i];
+                break;
+              }
             }
-            if (relativeParts[1] && relativeParts[2]) {
-              $scope.relative.count = parseInt(relativeParts[1], 10);
-              $scope.relative.unit = relativeParts[2];
-            } else {
-              let duration = moment.duration(moment().diff(dateMath.parse($scope.from)));
-              let units = _.pluck(_.clone($scope.relativeOptions).reverse(), 'value');
-              if ($scope.from.toString().split('/')[1]) $scope.relative.round = true;
-              for (let i = 0; i < units.length; i++) {
-                let as = duration.as(units[i]);
-                if (as > 1) {
-                  $scope.relative.count = Math.round(as);
-                  $scope.relative.unit = units[i];
-                  break;
-                }
+
+            for (i = 0; i < unitsFrom.length; i++) {
+              var asFrom = durationFrom.as(unitsFrom[i]);
```

Comment:
Since the current code breaks out on the first occurrance of as > 1, how would thi work with map? A map will not exit before iterating through all the elements in the list/array which is a) wasteful b)leads to additional code to figure out the first occurrance of a >1 value